### PR TITLE
Deep Imbalanced Regression implementation - mainly

### PR DESCRIPTION
**Pull request by 5uperpalo**
Main contributions:
* implementation of Label Distribution Smoothing, significantly adjusted code from https://github.com/YyzHarry/imbalanced-regression
  * WideDeepDataset object assigns weights to target values in regression objective, these weights are propagated to loss functions which use them for weighting
* implementation of Feature Distribution Smoothing, again significantly adjusted code from https://github.com/YyzHarry/imbalanced-regression
  * the FDS is possible to run only in deep tabular mode without wide part 

Minor contributions:
* adjusted library loss functions to be able to use weights from LDS
* added Focal-R loss functions from DeepImbalancedRegression repo/paper mentioned above
* adjusted dataset module - added new datasets + saved datasets as parquet, pyarrow, brotli
  * DataFrame.to_parquet(path=None, engine="auto", compression="brotli", index=False)
  * see related post: https://python.plainenglish.io/storing-pandas-98-faster-disk-reads-and-72-less-space-208e2e2be8bb
* run black to adjust notebooks + added typing where I noticed it is missing
**NOTES for FDS**
* model initialization:
  * widedeep.py,
    * line 197 - initialize FDS and dropout layer before prediction layer
    * line 334 - if the FDS is used the prediction layer must be separate, if the FDS is not used the prediction layer is part of the deeptabular model created by _add_pred_layer()
* model training:
  * prediction - widedeep.py,
    * line 225, if FDS is used then during the training the model outputs y_pred(using deeptabular()->FDS.smooth()->FDS_dropout()->pred_layer()) and also deeptab_features, ie. input that goes into FDS.smooth()
  * after prediction and optimization - trainer.py,
    * line 659 after every epoch, all encodings and labels are gathered using simple _fds_step() method (this does nothing special just gathers labels and inputs that go into the FDS layers)
  * in trainer.py I used
      *"if getattr(train_loader,"lds",False) == True:"
as it is "more robust", if I only use "if train_loader.lds" it created some issues in test_callbacks when data is directly put into fit method

### CODE REVIEW

**Comment by 5uperpalo** [comment]
this branch will be merged into jrzaurin/attention_mlp branch
